File: Leetcode/add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        head = ListNode(0)  # dummy head
        current = head

        while l1 or l2:
            if current is None:
                break
            
            v1 = l1.val if l1 else 0
            v2 = l2.val if l2 else 0
            sum_ = v1 + v2 + current.val

            current.val = sum_ % 10
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None

            current.next = ListNode(sum_ // 10) if any([sum_ >= 10, l1, l2]) else None
            current = current.next

        return head

File: Leetcode/add_two_numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode answer = new ListNode(0);
        ListNode answerRoot = null;

        int carry = 0;
        while(l1 != null || l2 != null || carry > 0) {
            ListNode next = new ListNode((l2 == null ? 0 : l2.val) + (l1 == null ? 0 : l1.val) + carry);
            carry = next.val / 10;
            next.val = next.val % 10;
            l1 = l1 != null ? l1.next : null;
            l2 = l2 != null ? l2.next : null;

            if(answer != null) {
                answer.next = next;
            }

            answer = next;

            if(answerRoot == null) {
                answerRoot = answer;
            }
        }

        if(answerRoot == null) {
            answerRoot = new ListNode(0);
        }

        return answerRoot;
    }
}
